### PR TITLE
Sema: fix OOB access in coerceTupleToStruct

### PR DESCRIPTION
**Pull request by wrongnull**
Closes https://github.com/ziglang/zig/issues/19529

### CODE REVIEW

**Comment by wrongnull** [comment]
not sure if I should add test for this

**Comment by Rexicon226** [comment]
Probably should add a test yes :)

File: src/Sema.zig [hunk 1]
```diff
@@ -32334,7 +32334,10 @@ fn coerceTupleToStruct(
                 anon_struct_type.names.get(ip)[tuple_field_index]
             else
                 try ip.getOrPutStringFmt(sema.gpa, "{d}", .{tuple_field_index}, .no_embedded_nulls),
-            .struct_type => ip.loadStructType(inst_ty.toIntern()).field_names.get(ip)[tuple_field_index],
+            .struct_type => if (ip.loadStructType(inst_ty.toIntern()).field_names.len > 0)
+                ip.loadStructType(inst_ty.toIntern()).field_names.get(ip)[tuple_field_index]
+            else
+                try ip.getOrPutStringFmt(sema.gpa, "{d}", .{tuple_field_index}, .no_embedded_nulls),
```
Comment (jacobly0) [inline src/Sema.zig:32340]:
```suggestion
            .struct_type => ip.loadStructType(inst_ty.toIntern()).fieldName(ip, tuple_field_index).unwrap() orelse
                try ip.getOrPutStringFmt(sema.gpa, "{d}", .{tuple_field_index}, .no_embedded_nulls),
```

**Review by jacobly0 — CHANGES_REQUESTED** [review]
> not sure if I should add test for this

The default answer to this question is always yes.

File: src/Sema.zig (hunk 1 above)
Comment (jacobly0) [reply src/Sema.zig:32340]: Although, really it should probably be more like:
```zig
	const maybe_field_name = switch (ip.indexToKey(inst_ty.toIntern())) {
	   .anon_struct_type => |anon_struct_type| anon_struct_type.fieldName(ip, tuple_field_index),
	   .struct_type => ip.loadStructType(inst_ty.toIntern()).fieldName(ip, tuple_field_index),
	   else => unreachable,
	};
	const struct_field_index = if (maybe_field_name.unwrap()) |field_name|
            try sema.structFieldIndex(block, struct_ty, field_name, field_src)
	else
	   tuple_field_index;
```

Comment (wrongnull) [reply src/Sema.zig]: done

**Review by jacobly0 — APPROVED** [review]
I was trying to avoid the extra call to `getOrPutStringFmt` in the `.anon_struct_type` case, but the new code is certainly reasonable enough that I won't hold up the merge over it.

**Comment by wrongnull** [comment]
Screenshot_20240603_214413

**Review by Vexu — APPROVED** [review]
Please use `Type.structFieldName` and also update the equivalent code in `coerceTupleToTuple` to use it.

I promise I'll merge this as soon as it passes the CI.

**Comment by wrongnull** [comment]
> Please use `Type.structFieldName` and also update the equivalent code in `coerceTupleToTuple` to use it.
> 
> I promise I'll merge this as soon as it passes the CI.

Done. However I'm not sure if pr won't reverted because of this https://github.com/ziglang/zig/pull/18071#issuecomment-1850740167

**Comment by Vexu** [comment]
> Done. However I'm not sure if pr won't reverted because of this [#18071 (comment)](https://github.com/ziglang/zig/pull/18071#issuecomment-1850740167)

Right, guess I'll leave it then.

**Comment by wrongnull** [comment]
I'll be waiting. No matter how long it takes

**Pull request merged**